---
title: "Data Cleaning"
format: html
---

In [1]:
import requests
import json
import re
import pandas as pd

I clean all the data here so I can begin to do analysis and try to build some useful models. I'll start by reading in a csv, selecting some columns, and changing the column names.

In [2]:
DP02cols=['NAME','DP02_0001E','DP02_0002E','DP02_0003E','DP02_0007E','DP02_0011E','DP02_0037E','DP02_0060E','DP02_0061E','DP02_0062E','DP02_0063E','DP02_0064E','DP02_0065E']
DP03cols=['NAME','DP03_0026PE','DP03_0005E','DP03_0062E','DP03_0063E','DP03_0093E','DP03_0094E']
DP04cols=['NAME','DP04_0002E','DP04_0001E','DP04_0037E','DP04_0047E','DP04_0134E']
DP05cols=['NAME','DP05_0001E','DP05_0004E','DP05_0018E','DP05_0037E','DP05_0038E','DP05_0039E','DP05_0044E','DP05_0073E']

In [3]:
DP0217=pd.read_csv('data/2017DP02.csv',skiprows=1)
DP0218=pd.read_csv('data/2018DP02.csv',skiprows=1)
DP0219=pd.read_csv('data/2019DP02.csv',skiprows=1)
DP0221=pd.read_csv('data/2021DP02.csv',skiprows=1)
DP0222=pd.read_csv('data/2022DP02.csv',skiprows=1)
DP0317 = pd.read_csv('data/2017DP03.csv', skiprows=1)
DP0318 = pd.read_csv('data/2018DP03.csv', skiprows=1)
DP0319 = pd.read_csv('data/2019DP03.csv', skiprows=1)
DP0321 = pd.read_csv('data/2021DP03.csv', skiprows=1)
DP0322 = pd.read_csv('data/2022DP03.csv', skiprows=1)
DP0417 = pd.read_csv('data/2017DP04.csv', skiprows=1)
DP0418 = pd.read_csv('data/2018DP04.csv', skiprows=1)
DP0419 = pd.read_csv('data/2019DP04.csv', skiprows=1)
DP0421 = pd.read_csv('data/2021DP04.csv', skiprows=1)
DP0422 = pd.read_csv('data/2022DP04.csv', skiprows=1)
DP0517 = pd.read_csv('data/2017DP05.csv', skiprows=1)
DP0518 = pd.read_csv('data/2018DP05.csv', skiprows=1)
DP0519 = pd.read_csv('data/2019DP05.csv', skiprows=1)
DP0521 = pd.read_csv('data/2021DP05.csv', skiprows=1)
DP0522 = pd.read_csv('data/2022DP05.csv', skiprows=1)

In [4]:
DP0217=DP0217[DP02cols]
DP0217['Year']=2017
DP0218=DP0218[DP02cols]
DP0218['Year']=2018
DP0219=DP0219[DP02cols]
DP0219['Year']=2019
DP0221=DP0221[DP02cols]
DP0221['Year']=2021
DP0222=DP0222[DP02cols]
DP0222['Year']=2022

In [5]:
DP0317 = DP0317[DP03cols]
DP0318 = DP0318[DP03cols]
DP0319 = DP0319[DP03cols]
DP0321 = DP0321[DP03cols]
DP0322 = DP0322[DP03cols]

In [6]:
DP0417 = DP0417[DP04cols]
DP0418 = DP0418[DP04cols]
DP0419 = DP0419[DP04cols]
DP0421 = DP0421[DP04cols]
DP0422 = DP0422[DP04cols]

In [7]:
DP0517 = DP0517[DP05cols]
DP0518 = DP0518[DP05cols]
DP0519 = DP0519[DP05cols]
DP0521 = DP0521[DP05cols]
DP0522 = DP0522[DP05cols]

In [8]:
df2017=pd.merge(DP0217,DP0317,on='NAME')
df2017=pd.merge(df2017,DP0417,on='NAME')
df2017=pd.merge(df2017,DP0517,on='NAME')

df2018=pd.merge(DP0218,DP0318,on='NAME')
df2018=pd.merge(df2018,DP0418,on='NAME')
df2018=pd.merge(df2018,DP0518,on='NAME')

df2019=pd.merge(DP0219,DP0319,on='NAME')
df2019=pd.merge(df2019,DP0419,on='NAME')
df2019=pd.merge(df2019,DP0519,on='NAME')


df2021=pd.merge(DP0221,DP0321,on='NAME')
df2021=pd.merge(df2021,DP0421,on='NAME')
df2021=pd.merge(df2021,DP0521,on='NAME')

df2022=pd.merge(DP0222,DP0322,on='NAME')
df2022=pd.merge(df2022,DP0422,on='NAME')
df2022=pd.merge(df2022,DP0522,on='NAME')

In [9]:
us_states = [
    "Alabama", "Alaska", "Arizona", "Arkansas", "California",
    "Colorado", "Connecticut", "Delaware", "Florida", "Georgia",
    "Hawaii", "Idaho", "Illinois", "Indiana", "Iowa",
    "Kansas", "Kentucky", "Louisiana", "Maine", "Maryland",
    "Massachusetts", "Michigan", "Minnesota", "Mississippi", "Missouri",
    "Montana", "Nebraska", "Nevada", "New Hampshire", "New Jersey",
    "New Mexico", "New York", "North Carolina", "North Dakota", "Ohio",
    "Oklahoma", "Oregon", "Pennsylvania", "Rhode Island", "South Carolina",
    "South Dakota", "Tennessee", "Texas", "Utah", "Vermont",
    "Virginia", "Washington", "West Virginia", "Wisconsin", "Wyoming",
    "District of Columbia"
]

In [10]:
feature_df=pd.DataFrame()

In [11]:
for state in us_states:
    row1=df2017[df2017['NAME']==state]
    row2=df2018[df2018['NAME']==state]
    row3=df2019[df2019['NAME']==state]
    row4=df2021[df2021['NAME']==state]
    row5=df2022[df2022['NAME']==state]
    df=pd.concat([row1,row2,row3,row4,row5])
    df=df.drop('NAME',axis=1)
    df=df.set_index('Year').pct_change().reset_index()
    df['RegionName']=state
    feature_df=pd.concat([feature_df,df])

In [12]:
feature_df

Year  DP02_0001E  DP02_0002E  DP02_0003E  DP02_0007E  DP02_0011E  \
0   2017         NaN         NaN         NaN         NaN         NaN   
1   2018    0.007341    0.009226   -0.008186    0.090413    0.007228   
2   2019    0.022851   -0.265678   -0.360479   -0.317095   -0.802204   
3   2021    0.036880    0.013836    0.056032   -0.010988    0.081890   
4   2022    0.024848    0.041137    0.053954   -0.058821    0.009307   
..   ...         ...         ...         ...         ...         ...   
0   2017         NaN         NaN         NaN         NaN         NaN   
1   2018    0.021320    0.018198    0.049960    0.151328    0.030406   
2   2019    0.014241   -0.355454   -0.490505   -0.365204   -0.869834   
3   2021    0.096015   -0.010890    0.155305   -0.343210    0.020759   
4   2022    0.023172    0.000929   -0.045818   -0.211063   -0.092319   

    DP02_0037E  DP02_0060E  DP02_0061E  DP02_0062E  ...  DP04_0134E  \
0          NaN         NaN         NaN         NaN  ...         NaN   
1     0.004783    0.022710   -0.002918    0.018174  ...    0.050667   
2     1.302990   -0.588360   -0.705200    0.439418  ...    0.024112   
3    -0.050374   -0.025077   -0.044056    0.038532  ...    0.066914   
4     0.106362   -0.070979   -0.062720   -0.020504  ...    0.060395   
..         ...         ...         ...         ...  ...         ...   
0          NaN         NaN         NaN         NaN  ...         NaN   
1     0.112103   -0.151330   -0.025084   -0.007725  ...    0.011341   
2     0.977101   -0.287304   -0.701942    0.298245  ...    0.057388   
3    -0.059821   -0.015675   -0.257760   -0.112117  ...    0.040549   
4     0.055245   -0.370224    0.098800   -0.089982  ...    0.104916   

    DP05_0001E  DP05_0004E  DP05_0018E  DP05_0037E  DP05_0038E  DP05_0039E  \
0          NaN         NaN         NaN         NaN         NaN         NaN   
1     0.002692    0.003198    0.010283   -0.001775   -0.000327   -0.123824   
2     0.003133   -0.006376    0.002545    0.005908    0.009572    0.054480   
3     0.027878    0.008556    0.010152   -0.013376   -0.010947    0.067268   
4     0.006829   -0.001060   -0.005025    0.006291   -0.002353   -0.017318   
..         ...         ...         ...         ...         ...         ...   
0          NaN         NaN         NaN         NaN         NaN         NaN   
1     0.012224    0.006652   -0.002941    0.042915    0.004893    0.143195   
2     0.004689   -0.008811    0.011799    0.011158    0.002899   -0.023810   
3    -0.050583    0.010000    0.014577   -0.154024   -0.097916   -0.320785   
4     0.002616   -0.001100    0.002874    0.016235   -0.023242    1.443404   

    DP05_0044E  DP05_0073E            RegionName  
0          NaN         NaN               Alabama  
1    -0.027097    0.217687               Alabama  
2     0.015884   -0.178100               Alabama  
3     0.040587   -0.005625               Alabama  
4     0.146484   10.243876               Alabama  
..         ...         ...                   ...  
0          NaN         NaN  District of Columbia  
1    -0.033720    0.598440  District of Columbia  
2     0.042995    0.392450  District of Columbia  
3    -0.035861   -0.005959  District of Columbia  
4     0.009714    8.276008  District of Columbia  

[255 rows x 33 columns]

In [13]:
feature_df=feature_df.dropna()

In [14]:
feature_df['Year']=feature_df['Year'].astype(str)

C:\Users\npeno\AppData\Local\Temp\ipykernel_648\58169971.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  feature_df['Year']=feature_df['Year'].astype(str)


In [15]:
feature_df

Year  DP02_0001E  DP02_0002E  DP02_0003E  DP02_0007E  DP02_0011E  \
1   2018    0.007341    0.009226   -0.008186    0.090413    0.007228   
2   2019    0.022851   -0.265678   -0.360479   -0.317095   -0.802204   
3   2021    0.036880    0.013836    0.056032   -0.010988    0.081890   
4   2022    0.024848    0.041137    0.053954   -0.058821    0.009307   
1   2018    0.015195    0.000753   -0.043082   -0.078405    0.043289   
..   ...         ...         ...         ...         ...         ...   
4   2022    0.002299   -0.012945   -0.040358   -0.228287   -0.235682   
1   2018    0.021320    0.018198    0.049960    0.151328    0.030406   
2   2019    0.014241   -0.355454   -0.490505   -0.365204   -0.869834   
3   2021    0.096015   -0.010890    0.155305   -0.343210    0.020759   
4   2022    0.023172    0.000929   -0.045818   -0.211063   -0.092319   

    DP02_0037E  DP02_0060E  DP02_0061E  DP02_0062E  ...  DP04_0134E  \
1     0.004783    0.022710   -0.002918    0.018174  ...    0.050667   
2     1.302990   -0.588360   -0.705200    0.439418  ...    0.024112   
3    -0.050374   -0.025077   -0.044056    0.038532  ...    0.066914   
4     0.106362   -0.070979   -0.062720   -0.020504  ...    0.060395   
1     0.942898   -0.208181    0.031255    0.017837  ...   -0.019983   
..         ...         ...         ...         ...  ...         ...   
4     0.039527    0.185354   -0.089400    0.011854  ...    0.006749   
1     0.112103   -0.151330   -0.025084   -0.007725  ...    0.011341   
2     0.977101   -0.287304   -0.701942    0.298245  ...    0.057388   
3    -0.059821   -0.015675   -0.257760   -0.112117  ...    0.040549   
4     0.055245   -0.370224    0.098800   -0.089982  ...    0.104916   

    DP05_0001E  DP05_0004E  DP05_0018E  DP05_0037E  DP05_0038E  DP05_0039E  \
1     0.002692    0.003198    0.010283   -0.001775   -0.000327   -0.123824   
2     0.003133   -0.006376    0.002545    0.005908    0.009572    0.054480   
3     0.027878    0.008556    0.010152   -0.013376   -0.010947    0.067268   
4     0.006829   -0.001060   -0.005025    0.006291   -0.002353   -0.017318   
1    -0.003186   -0.009174    0.011594   -0.000634    0.139775    0.012942   
..         ...         ...         ...         ...         ...         ...   
4     0.004454    0.003820    0.002564    0.004535   -0.387319    0.073581   
1     0.012224    0.006652   -0.002941    0.042915    0.004893    0.143195   
2     0.004689   -0.008811    0.011799    0.011158    0.002899   -0.023810   
3    -0.050583    0.010000    0.014577   -0.154024   -0.097916   -0.320785   
4     0.002616   -0.001100    0.002874    0.016235   -0.023242    1.443404   

    DP05_0044E  DP05_0073E            RegionName  
1    -0.027097    0.217687               Alabama  
2     0.015884   -0.178100               Alabama  
3     0.040587   -0.005625               Alabama  
4     0.146484   10.243876               Alabama  
1    -0.060828    0.054155                Alaska  
..         ...         ...                   ...  
4    -0.233665   42.074760               Wyoming  
1    -0.033720    0.598440  District of Columbia  
2     0.042995    0.392450  District of Columbia  
3    -0.035861   -0.005959  District of Columbia  
4     0.009714    8.276008  District of Columbia  

[204 rows x 33 columns]

In [16]:
zillow=pd.read_csv('data/Zillow.csv')

In [17]:
zillowlong=zillow.melt(id_vars=['RegionName'],var_name='Year',value_name='Typical Home Value')

In [18]:
zillowlong

RegionName  Year  Typical Home Value
0              California  2017       498316.221451
1                   Texas  2017       192963.504413
2                 Florida  2017       221751.216789
3                New York  2017       316846.572473
4            Pennsylvania  2017       174153.095534
..                    ...   ...                 ...
250          North Dakota  2022       254075.792578
251                Alaska  2022       350130.429150
252  District of Columbia  2022       618945.155475
253               Vermont  2022       368730.928380
254               Wyoming  2022       331644.235988

[255 rows x 3 columns]

In [19]:
target_df=pd.DataFrame()

In [20]:
for state in us_states:
    row1=zillowlong[(zillowlong['RegionName']==state)&(zillowlong['Year']=='2017')]
    row2=zillowlong[(zillowlong['RegionName']==state)&(zillowlong['Year']=='2018')]
    row3=zillowlong[(zillowlong['RegionName']==state)&(zillowlong['Year']=='2019')]
    row4=zillowlong[(zillowlong['RegionName']==state)&(zillowlong['Year']=='2021')]
    row5=zillowlong[(zillowlong['RegionName']==state)&(zillowlong['Year']=='2022')]
    df=pd.concat([row1,row2,row3,row4,row5])
    df=df.drop('RegionName',axis=1)
    df=df.set_index('Year').pct_change().reset_index()
    df['RegionName']=state
    target_df=pd.concat([target_df,df])

In [21]:
target_df

Year  Typical Home Value            RegionName
0   2017                 NaN               Alabama
1   2018            0.038688               Alabama
2   2019            0.070360               Alabama
3   2021            0.263465               Alabama
4   2022            0.076587               Alabama
..   ...                 ...                   ...
0   2017                 NaN  District of Columbia
1   2018            0.055570  District of Columbia
2   2019            0.017659  District of Columbia
3   2021            0.097177  District of Columbia
4   2022           -0.009417  District of Columbia

[255 rows x 3 columns]

In [22]:
target_df=target_df.dropna()

In [23]:
final_data=pd.merge(feature_df,target_df,on=['RegionName','Year'])

In [24]:
final_data

Year  DP02_0001E  DP02_0002E  DP02_0003E  DP02_0007E  DP02_0011E  \
0    2018    0.007341    0.009226   -0.008186    0.090413    0.007228   
1    2019    0.022851   -0.265678   -0.360479   -0.317095   -0.802204   
2    2021    0.036880    0.013836    0.056032   -0.010988    0.081890   
3    2022    0.024848    0.041137    0.053954   -0.058821    0.009307   
4    2018    0.015195    0.000753   -0.043082   -0.078405    0.043289   
..    ...         ...         ...         ...         ...         ...   
199  2022    0.002299   -0.012945   -0.040358   -0.228287   -0.235682   
200  2018    0.021320    0.018198    0.049960    0.151328    0.030406   
201  2019    0.014241   -0.355454   -0.490505   -0.365204   -0.869834   
202  2021    0.096015   -0.010890    0.155305   -0.343210    0.020759   
203  2022    0.023172    0.000929   -0.045818   -0.211063   -0.092319   

     DP02_0037E  DP02_0060E  DP02_0061E  DP02_0062E  ...  DP05_0001E  \
0      0.004783    0.022710   -0.002918    0.018174  ...    0.002692   
1      1.302990   -0.588360   -0.705200    0.439418  ...    0.003133   
2     -0.050374   -0.025077   -0.044056    0.038532  ...    0.027878   
3      0.106362   -0.070979   -0.062720   -0.020504  ...    0.006829   
4      0.942898   -0.208181    0.031255    0.017837  ...   -0.003186   
..          ...         ...         ...         ...  ...         ...   
199    0.039527    0.185354   -0.089400    0.011854  ...    0.004454   
200    0.112103   -0.151330   -0.025084   -0.007725  ...    0.012224   
201    0.977101   -0.287304   -0.701942    0.298245  ...    0.004689   
202   -0.059821   -0.015675   -0.257760   -0.112117  ...   -0.050583   
203    0.055245   -0.370224    0.098800   -0.089982  ...    0.002616   

     DP05_0004E  DP05_0018E  DP05_0037E  DP05_0038E  DP05_0039E  DP05_0044E  \
0      0.003198    0.010283   -0.001775   -0.000327   -0.123824   -0.027097   
1     -0.006376    0.002545    0.005908    0.009572    0.054480    0.015884   
2      0.008556    0.010152   -0.013376   -0.010947    0.067268    0.040587   
3     -0.001060   -0.005025    0.006291   -0.002353   -0.017318    0.146484   
4     -0.009174    0.011594   -0.000634    0.139775    0.012942   -0.060828   
..          ...         ...         ...         ...         ...         ...   
199    0.003820    0.002564    0.004535   -0.387319    0.073581   -0.233665   
200    0.006652   -0.002941    0.042915    0.004893    0.143195   -0.033720   
201   -0.008811    0.011799    0.011158    0.002899   -0.023810    0.042995   
202    0.010000    0.014577   -0.154024   -0.097916   -0.320785   -0.035861   
203   -0.001100    0.002874    0.016235   -0.023242    1.443404    0.009714   

     DP05_0073E            RegionName  Typical Home Value  
0      0.217687               Alabama            0.038688  
1     -0.178100               Alabama            0.070360  
2     -0.005625               Alabama            0.263465  
3     10.243876               Alabama            0.076587  
4      0.054155                Alaska            0.005789  
..          ...                   ...                 ...  
199   42.074760               Wyoming            0.083624  
200    0.598440  District of Columbia            0.055570  
201    0.392450  District of Columbia            0.017659  
202   -0.005959  District of Columbia            0.097177  
203    8.276008  District of Columbia           -0.009417  

[204 rows x 34 columns]

In [25]:
final_data.to_csv('./data/RecordData.csv', index=False)

In [193]:
DP02colss=['DP02_0001E','DP02_0002E','DP02_0003E','DP02_0007E','DP02_0011E','DP02_0037E','DP02_0060E','DP02_0061E','DP02_0062E','DP02_0063E','DP02_0064E','DP02_0065E']
DP03colss=['DP03_0026PE','DP03_0005E','DP03_0062E','DP03_0063E','DP03_0093E','DP03_0094E']
DP04colss=['DP04_0002E','DP04_0001E','DP04_0037E','DP04_0047E','DP04_0134E']
DP05colss=['DP05_0001E','DP05_0004E','DP05_0018E','DP05_0037E','DP05_0038E','DP05_0039E','DP05_0044E','DP05_0073E']

In [194]:
for name in DP02colss:
    print(feature_df[name].max())

0.0960146791508043
0.08325965443347161
0.21027912686129047
0.5718313168095854
0.31098100619511193
4.71179314694135
0.5836723372781065
0.2331338411316648
1.16632039564257
3.7135790077430455
0.20096335286522993
1.8600379486638907


In [195]:
for name in DP03colss:
    print(feature_df[name].max())

0.08159905963183367
0.9855738883335392
0.1558987776230465
0.13903133903133913
0.18793842369265668
0.19923024054982807


In [196]:
for name in DP04colss:
    print(feature_df[name].max())

0.0960146791508043
0.10739311182290723
0.03921568627450989
0.1304236996210817
0.17613636363636354


In [197]:
for name in DP05colss:
    print(feature_df[name].max())

0.06371228802533757
0.024340770791075217
0.026666666666666616
0.04291548345376328
0.9405508590128171
1.598558282208589
0.3942459396751741
85.27326007326008


4.71179314694135
85.27326007326008

We may need to drop some columns that are producing outliers.